File: src/util/logger.c
#include <stdlib.h>
#include <stdarg.h>
#include <stdio.h>
#include "logger.h"

#define PRINT_PREFIX   WOLF_ANSI_RESET  "[ ]: " WOLF_ANSI_RESET
#define NOTICE_PREFIX  WOLF_ANSI_CYAN   "[-]: " WOLF_ANSI_RESET
#define WARN_PREFIX    WOLF_ANSI_YELLOW "[*]: " WOLF_ANSI_RESET
#define ERROR_PREFIX   WOLF_ANSI_RED    "[!]: " WOLF_ANSI_RESET
#define SUCCESS_PREFIX WOLF_ANSI_GREEN  "[^]: " WOLF_ANSI_RESET

#define PRINT_TEMPLATE(target, exec) va_list args;                       \
                                           va_start(args, fmt);                \
                                           FILE* print_stream = (FILE*)target; \
                                           exec                                \
                                           va_end(args)

static FILE* print_stream_ptr = NULL;
static FILE* error_stream_ptr = NULL;



void* wolf_print_stream() {
    if(print_stream_ptr==NULL) {
        print_stream_ptr = stdout;
    }
    return print_stream_ptr;
}
void* wolf_error_stream() {
    if(error_stream_ptr==NULL) {
        error_stream_ptr = stderr;
    }
    return error_stream_ptr;
}
bool wolf_retarget_print(void* target) {
    if(target==NULL) {
        return false;
    } else {
        print_stream_ptr = (FILE*)target;
        return true;
    }
}
bool wolf_retarget_error(void* target) {
    if(target==NULL) {
        return false;
    } else {
        error_stream_ptr = (FILE*)target;
        return true;
    }
}

void wolf_print    (const char* fmt, ...) {
    PRINT_TEMPLATE(wolf_print_stream(),

    fprintf(print_stream,PRINT_PREFIX);
    vfprintf(print_stream, fmt, args);

    );
}
void wolf_println  (const char* fmt, ...) {
    PRINT_TEMPLATE(wolf_print_stream(),

    fprintf(print_stream,PRINT_PREFIX);
    vfprintf(print_stream, fmt, args);
    fprintf(print_stream, "\n");

    );
}
void wolf_notice   (const char* fmt, ...) {
    PRINT_TEMPLATE(wolf_print_stream(),

    fprintf(print_stream,NOTICE_PREFIX);
    vfprintf(print_stream, fmt, args);
    
    );
}
void wolf_noticeln (const char* fmt, ...) {
    PRINT_TEMPLATE(wolf_print_stream(),

    fprintf(print_stream,NOTICE_PREFIX);
    vfprintf(print_stream, fmt, args);
    fprintf(print_stream, "\n");
    
    );
}
void wolf_warn     (const char* fmt, ...) {
    PRINT_TEMPLATE(wolf_error_stream(),

    fprintf(print_stream,WARN_PREFIX);
    vfprintf(print_stream, fmt, args);
    
    );
}
void wolf_warnln   (const char* fmt, ...) { 
    PRINT_TEMPLATE(wolf_error_stream(),

    fprintf(print_stream,WARN_PREFIX);
    vfprintf(print_stream, fmt, args);
    fprintf(print_stream, "\n");
    
    );
}
void wolf_error    (const char* fmt, ...) {
    PRINT_TEMPLATE(wolf_error_stream(),

    fprintf(print_stream,ERROR_PREFIX);
    vfprintf(print_stream, fmt, args);
    
    );
}
void wolf_errorln  (const char* fmt, ...) {
    PRINT_TEMPLATE(wolf_error_stream(),

    fprintf(print_stream,ERROR_PREFIX);
    vfprintf(print_stream, fmt, args);
    fprintf(print_stream, "\n");
    
    );
}
void wolf_success (const char* fmt, ...) { 
    PRINT_TEMPLATE(wolf_print_stream(),

    fprintf(print_stream,SUCCESS_PREFIX);
    vfprintf(print_stream, fmt, args);
    
    );
}
void wolf_successln(const char* fmt, ...) {
    PRINT_TEMPLATE(wolf_print_stream(),

    fprintf(print_stream,SUCCESS_PREFIX);
    vfprintf(print_stream, fmt, args);
    fprintf(print_stream, "\n");
    
    );
}

void wolf_print_raw    (const char* fmt, ...) {
    PRINT_TEMPLATE(wolf_print_stream(),

    vfprintf(print_stream, fmt, args);

    );
}

void wolf_println_raw  (const char* fmt, ...) {
    PRINT_TEMPLATE(wolf_print_stream(),

    vfprintf(print_stream, fmt, args);
    fprintf (print_stream, "\n");

    );
}

void wolf_error_raw   (const char* fmt, ...) {
    PRINT_TEMPLATE(wolf_error_stream(),

    vfprintf(print_stream, fmt, args);

    );
}
void wolf_errorln_raw (const char* fmt, ...) {
    PRINT_TEMPLATE(wolf_error_stream(),

    vfprintf(print_stream, fmt, args);
    fprintf (print_stream, "\n");

    );
}

File: src/wolf/util/logger.nim
##
##
##     Five types of print function each with a newline varient:
##         print
##         notice
##         warn
##         error
##         success
##
##

{.push header: "\"util/logger.h\"".}

when not defined(WOLF_NO_ANSI_COLOUR):
  const
    ANSI_RED* = "\e[31m"
    ANSI_GREEN* = "\e[32m"
    ANSI_YELLOW* = "\e[33m"
    ANSI_BLUE* = "\e[34m"
    ANSI_MAGENTA* = "\e[35m"
    ANSI_CYAN* = "\e[36m"
    ANSI_RESET* = "\e[0m"
else:
  const
    ANSI_RED* = ""
    ANSI_GREEN* = ""
    ANSI_YELLOW* = ""
    ANSI_BLUE* = ""
    ANSI_MAGENTA* = ""
    ANSI_CYAN* = ""
    ANSI_RESET* = ""


import
  common

##  returns a FILE*
proc printStream*(): pointer {.importc: "wolf_print_stream".}

##  returns a FILE*
proc errorStream*(): pointer {.importc: "wolf_error_stream".}

##   stream must be a FILE*!
proc retargetPrint*(stream: pointer): bool {.importc: "wolf_retarget_print".}

##  stream must be a FILE*!
proc retargetError*(stream: pointer): bool {.importc: "wolf_retarget_error".}

##  outputs to print stream, prefixed with '[ ]: '
proc rint*(fmt: cstring) {.varargs, importc: "wolf_print".}

##  outputs to print stream, prefixed with '[ ]: ' and newlined
proc println*(fmt: cstring) {.varargs, importc: "wolf_println".}

##  outputs to print stream, prefixed with '[-]: '
proc notice*(fmt: cstring) {.varargs, importc: "wolf_notice".}

##  outputs to print stream, prefixed with '[-]: ' and newlined
proc noticeln*(fmt: cstring) {.varargs, importc: "wolf_noticeln".}

##  outputs to error stream, prefixed with '[*]: '
proc warn*(fmt: cstring) {.varargs, importc: "wolf_warn".}

##  outputs to error stream, prefixed with '[*]: ' and newlined
proc warnln*(fmt: cstring) {.varargs, importc: "wolf_warnln".}

##  outputs to error stream, prefixed with '[!]: '
proc error*(fmt: cstring) {.varargs, importc: "wolf_error".}

##  outputs to error stream, prefixed with '[!]: ' and newlined
proc errorln*(fmt: cstring) {.varargs, importc: "wolf_errorln".}

##  outputs to print stream, prefixed with '[^]: '
proc success*(fmt: cstring) {.varargs, importc: "wolf_success".}

##  outputs to print stream, prefixed with '[^]: ' and newlined
proc successln*(fmt: cstring) {.varargs, importc: "wolf_successln".}

##  outputs to print stream
proc printRaw*(fmt: cstring) {.varargs, importc: "wolf_print_raw".}

##  outputs to print stream and newlined
proc printlnRaw*(fmt: cstring) {.varargs, importc: "wolf_println_raw".}

##  outputs to error stream
proc errorRaw*(fmt: cstring) {.varargs, importc: "wolf_error_raw".}

##  outputs to print stream and newlined
proc errorlnRaw*(fmt: cstring) {.varargs, importc: "wolf_errorln_raw".}

{.pop.}